BrightSpots/rctab-docs
# Create a print page

## What this does:
* creates a print page that consolidates the entire PDF (for now, this is semi-hidden and requires you to know the path of the print page)
* updates dependencies and removes some unused dependencies related to other attempts to get a PDF export from MKDocs.
* minor fix to the styling of some internal links
* removes links from PDF print page
* adds instructions for generating PDF to the `README.md` file

## How to use test this PR:

1. Follow the README.md's instructions, including building and running mkdocs.
2. Then navigate to the print page using the Chrome browser. If you are running this locally, it should be something like http://localhost:8000/tdp/print-page.html
3. Then use your browser's/OS's print to PDF feature to get a PDF copy of the TDP
4. Open the PDF and check the following:
    - [ ] all sections are in order
    - [ ] no sections are missing
    - [ ] no sections are duplicated
    - [ ] no content is cut off the page
    - [ ] no content looks really wacky or strange compared to what is in MKDocs on the `main` branch of this repo (i.e. https://github.com/BrightSpots/rctab-docs/)
    - [ ] all links are stripped from PDF

## Still need to:
* ✅ do sanity check to make sure nothing is missing
* ✅  do sanity check to make sure that rendering/formatting is alright
* ✅  remove hyperlinks from print page
* ✅ make initial title page look prettier
* ✅  explore what header and footer options we have (we decided to defer this and maybe handle it in something like Adobe Acrobat)
* ~explore what options we have in terms of altering the page's margins~ (might not be needed; but we can use [CSS media queries](https://developer.mozilla.org/en-US/docs/Web/CSS/CSS_media_queries/Printing) if needed)
* 🚧 decide if we want a table of contents separate from the document abstracts **(needs further discussion)**

## Advantages:
* relatively easy implementation (especially if we are willing to accept some of the drawbacks)
* we can manually define what is in the document
* we have complete control over the table of contents
* we have complete control over section numbering
* we own the solution

## Drawbacks and Tradeoffs:
* internal links to other sections or content in other sections will be broken (see this URL for a potential solution: https://github.com/timvink/mkdocs-print-site-plugin/blob/master/mkdocs_print_site_plugin/urls.py)
* some items (one fixed set of tables that we knew about so far) may get cut off. Tends to be more likely to occur when we are manually writing HTML and CSS rather than just markdown. (Solution is to use custom [CSS media queries for printing](https://developer.mozilla.org/en-US/docs/Web/CSS/CSS_media_queries/Printing))
* any time we make edits to the TDP, we will have to sanity check both the HTML and PDF output. This will be more critical when we are introducing or changing HTML manually in the markdown files (which is sometimes required for more customized content layouts).
* different browsers (especially those with different rendering engines) may render the print version of an HTML site differently.
* custom headers and footers will not be easy to implement. It *might* be possible using some custom CSS and accessing the Chrome APIs (https://chromedevtools.github.io/devtools-protocol/tot/Page/#method-printToPDF). But this wouldn't work on all browsers
* can be a bit difficult to get manual CSS changes in MKDocs to have correct specificity and importance to be rendered.
* will require manual changes when:
    - new sections are added
    - sections are re-ordered
    - sections are removed
    - the paths to sections change
    - the table of contents changes
    - any time any content changes if we include page numbers in the table of contents
* styling quirks
    - sometimes table headers or sub-section headers are at the bottom of a page -- we should look into solutions
    - tables that lean heavily into MKDocs material repeat the borders a the top of the page -- we should replicate that with the custom HTML tables
    - sometimes due to image sizes and space limitations, an image and the heading or description text accompanying the image are on different pages -- This *might* be able to be fixed using HTML's captions figure. However that would have the trade off of leaning more into HTML (and potentially CSS) and less into MKDocs. It would take time to transition images using MKDocs styling to HMTL and then test those changes.
    - appendices are not currently migrated to MKDocs. For now, we would have to manually append those to the PDF, until those are added to MKDocs. We would also have to figure out a way to distribute those to amended PDFs to stakeholders.
    - the default headers and footers applied by Chrome are quite ugly. You can turn them off (and loose page numbering). Not quite sure how to customize them without digging deep in some Chrome APIs (https://chromedevtools.github.io/devtools-protocol/tot/Page/#method-printToPDF).

### Visual examples

| Description | Screenshot |
|------------|-------------|
| Different rendering by different browsers (Chrome on left; Firefox on right) | <img width="1438" alt="Screenshot 2024-09-06 at 12 41 38" src="https://github.com/user-attachments/assets/5238e57e-9551-4adb-82e9-4f3b404b2c4e"> |
| Headers on MKDocs markdown table repeat at the start of a new page (this is good) | <img width="648" alt="Screenshot 2024-09-06 at 12 24 40" src="https://github.com/user-attachments/assets/3373ff32-e9fb-4a84-929c-f22335d2a02a"> |
| Headers on custom HTML table don't repeat at the start of a new page | <img width="858" alt="Screenshot 2024-09-06 at 12 31 04" src="https://github.com/user-attachments/assets/c2883790-c4b1-458a-938e-e19f64be113e"> |
| Table headers at bottom of page | <img width="840" alt="Screenshot 2024-09-06 at 11 54 26" src="https://github.com/user-attachments/assets/4b48453c-e440-41ce-9818-07e99c75e581"> |
| Sub-section header at bottom of page | <img width="820" alt="Screenshot 2024-09-06 at 12 04 50" src="https://github.com/user-attachments/assets/dad9d667-db64-4557-a0eb-649fe9dc658b"> |
| Image description and actual image on different pages | <img width="647" alt="Screenshot 2024-09-06 at 12 13 14" src="https://github.com/user-attachments/assets/d2a4fe26-b6fb-43b7-8ac6-0ff6d0711108"> |
| Awkward extra space due to image sizes | <img width="224" alt="Screenshot 2024-09-06 at 12 11 07" src="https://github.com/user-attachments/assets/3f603e92-8164-4356-b86f-4bfd74b86d48"> |
| Ugly headers and footers | <img width="840" alt="Screenshot 2024-09-06 at 11 48 20" src="https://github.com/user-attachments/assets/c4e794aa-2be4-4564-a17d-12dfb58f31fd"> |

## Mitigation:

Some of these issues may be fixed with various amounts of work and effort. Other issues may be accepted as reasonable tradeoffs. When we accept trade offs, we should clearly communicate the limitations to stakeholders.
